backend/src/worker.rs

```diff
@@ -1351,175 +1360,261 @@ async fn handle_child(
+        let mut interval = interval_skipping_missed(cancel_check_interval).boxed();
+        while let Some(_) = interval.next().await {
+            if sqlx::query_scalar!("SELECT canceled FROM queue WHERE id = $1", job_id)
+                .fetch_optional(&db)
+                .await
+                .map(|v| Some(true) == v)
+                .unwrap_or_else(|err| {
+                    tracing::error!(%job_id, %err, "error checking cancelation for job {job_id}: {err}");
+                    false
+                })
+            {
+                break;
             }
-            _ = async move {
-                while !done2.load(Ordering::Relaxed) {
-                    tokio::time::sleep(Duration::from_secs(1)).await;
+        }
+    };
+
+    /* a future that completes when the child process exits */
+    let wait_on_child = async {
+        let db = db.clone();
+
+        let timed_out = tokio::select! {
+            biased;
+            result = child.wait() => return result.map(Some),
+            _ = too_many_logs.changed() => false,
+            _ = cancel_check => false,
+            _ = sleep(timeout) => true,
+        };
+
+        let set_reason = async {
+            if timed_out {
+                if let Err(err) = sqlx::query(
+                    r#"
+                       UPDATE queue
+                          SET canceled = true
+                            , canceled_by = 'timeout',
+                            , canceled_reason = $1
+                        WHERE id = $2
+                    r"#,
+                )
+                .bind(format!("duration > {}", timeout.as_secs()))
+                .bind(job_id)
+                .execute(&db)
+                .await
+                {
+                    tracing::error!(%job_id, %err, "error setting cancelation reason for job {job_id}: {err}");
                 }
-            } => {
-                child.kill().await?;
-                return Err(Error::ExecutionErr("execution interrupted".to_string()).into())
             }
         };
-        r
-    });
-
-    let (tx, mut rx) = mpsc::channel::<String>(100);
-
-    tokio::spawn(async move {
-        while !done4.load(Ordering::Relaxed) {
-            let send = tokio::select! {
-                Ok(Some(out)) = reader.next_line() => {
-                    if out.len() > MAX_LOG_SIZE as usize {
-                        tracing::info!("Line is too big");
-                        let _ = tx.send(format!("Line is too big")).await;
-                        done4.store(true, Ordering::Relaxed);
-                        break;
-                    } else {
-                        tx.send(out).await
+
+        /* send SIGKILL and reap child process */
+        let (_, kill) = future::join(set_reason, child.kill()).await;
+        kill.map(|()| None)
+    };
+
+    /* a future that reads output from the child and appends to the database */
+    /* this whole section is kind of a mess and could use some love */
```

Comment:
Remove this kind of language but list what could be improved if you want

Reply:
When you gave me feedback earlier you said you were surprised that I wrote this the way because it wasn't very clear or something. I shared the general sentiment. This comment was just to be clear about that so that it would be less surprising to readers.